Add DISTANCE_SENSOR support

khancyr opened the pull request:
This address https://github.com/ArduPilot/ardupilot/issues/4750 .
Tested on SITL and with mavros.

To test with mavros : in apm_config.yaml add `distance_sensor:
  hrlv_ez4_pub:
    id: 0
    frame_id: "sonar"
    orientation: PITCH_270 # RPY:{180.0, 0.0, 0.0}
    field_of_view: 0.0  # XXX TODO
    send_tf: true
    sensor_position: {x:  0.0, y:  0.0, z:  -0.1}`
in apm_pluginlists.yaml , remove distance_sensor from backlist
and add a sonar on sitl. Loading gazebo-iris.parm works.

rmackay9 commented:
thanks for this change, I'm keen to add it.  A few things though:

- no need for a new parameter for the sensor type I think.  I think we should always hard-code it to be Lidar or we should add a function to each backend so each driver can say what type of range finder it is.
- could the send_distancesensor be moved into the GCS_MAVLink library?  A bit similar to the "send_battery_status".

khancyr commented:
I thought about only hardcode value to. But for analog and pwm type you lost an information .
I can move the send function to GCS_MAVLink.

OXINARF commented on ArduCopter/GCS_Mavlink.cpp:
```diff
@@ -219,6 +219,25 @@ void NOINLINE Copter::send_rangefinder(mavlink_channel_t chan)
+
+void NOINLINE Copter::send_distance_sensor(mavlink_channel_t chan)
+{
+    for (uint8_t i = 0; i < RANGEFINDER_MAX_INSTANCES; i++) {
```
I understand this is debatable, but I think this should be doing the same as the rangefinder message - only send the one looking down; when you have others you'll most likely have the proximity feature enabled and the proximity method will already send messages.

rmackay9 replied:
yes, I agree with Francisco that for Copter we should only send the downward facing one for now at least because the send_proximity will send the rest.  The only way this isn't the case is if the proximity library isn't being used.. which is possible but probably rare.

OXINARF commented on ArduCopter/GCS_Mavlink.cpp:
```diff
@@ -475,6 +494,8 @@ bool GCS_MAVLINK_Copter::try_send_message(enum ap_message id)
         copter.send_rangefinder(chan);
 #endif
         CHECK_PAYLOAD_SIZE(DISTANCE_SENSOR);
+        copter.send_distance_sensor(chan);
```
This should be inside the if/endif.

khancyr replied:
done!

OXINARF commented on ArduCopter/GCS_Mavlink.cpp:
```diff
@@ -219,6 +219,25 @@ void NOINLINE Copter::send_rangefinder(mavlink_channel_t chan)
+                rangefinder.get_sensor_type(i),  // type from MAV_DISTANCE_SENSOR enum
+                i,                               // onboard ID of the sensor == instance ?
+                rangefinder.get_orientation(i),  // direction the sensor faces from MAV_SENSOR_ORIENTATION enum
+                1);                              // Measurement covariance in centimeters, 0 for unknown / invalid readings
```
Shouldn't this be 0?

khancyr replied:
Corrected

vooon commented:
@khancyr 0.19 already have configuration for APM: [apm_config.yaml L115-L126](https://github.com/mavlink/mavros/blob/master/mavros/launch/apm_config.yaml#L115-L126)

vooon commented:
Also IMHO instead of `send_tf: true` better to add base_link → lidar (or whatever set in `frame_id`) to URDF.
```xml
  <joint name="base_to_lidar" type="fixed">
    <parent link="base_link" />
    <child link="lidar" />
    <origin xyz="0.05 0 -0.03" rpy="0 ${pi} 0" /> <!-- meters / radians -->
  </joint>
```

khancyr commented:
@rmackay9 @OXINARF OK, rebased, and changes done.
I rework to remove the addition of parameter, the sensor type is now write into the backend. @OXINARF the mavlink enum cover all cases correctly ! (Or ToF sensor is missing?) But in case of PWM or analog rangefinder connection the sensor could be all type (ultrasound, IR or laser). It now fixed to Laser type by default.
The Distance_sensor sending function is now in gcs_common !
On copter, only the downard looking sensors will be send with DISTANCE_SENSOR like for rangefinder msg. I update the way proximity send the message otherwise all messages were mixed into the id 0 ...
The last thing to check is if proximity is enable, and got a downward looking sensors, are the id mixed or not ? I didn't manage to check this on SITL.

khancyr commented:
Only checked on SITL with copter for now

khancyr commented:
@jaxxzer Hello, can you have a quick look. I need your input to know if by default on Sub, DISTANCE_SENSOR send distance from all sensors or do like on copter and only send the downward looking distance ?

khancyr commented on libraries/AP_RangeFinder/AP_RangeFinder_LeddarOne.cpp:
```diff
@@ -27,7 +27,7 @@ extern const AP_HAL::HAL& hal;
                                                                RangeFinder::RangeFinder_State &_state,
                                                                AP_SerialManager &serial_manager) :
-    AP_RangeFinder_Backend(_ranger, instance, _state)
+    AP_RangeFinder_Backend(_ranger, instance, _state, MAV_DISTANCE_SENSOR_INFRARED)
```
they do ToF calculation with IR led array with 3deg beam, so I though it is more Infrared. I just look on ROS and they use laserscan msg. So I will change for laser

khancyr replied:
done

khancyr commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -219,6 +219,30 @@ bool GCS_MAVLINK::send_battery_status(const AP_BattMonitor &battery) const
+void GCS_MAVLINK::send_distance_sensor(const RangeFinder &rangefinder) const
+{
+    for (uint8_t i = 0; i < RANGEFINDER_MAX_INSTANCES; i++) {
+        send_distance_sensor(rangefinder, i);
```
ok!

khancyr replied:
done

rmackay9 commented:
sending the proximity ids from 10 seems like a reasonable work around for now anyway.

khancyr commented:
@rmackay9 So we start proximity id at 10 and upward proximity got id 19 right?

khancyr commented on libraries/AP_RangeFinder/RangeFinder.h:
```diff
@@ -154,6 +154,8 @@ class RangeFinder
     }
     int16_t ground_clearance_cm_orient(enum Rotation orientation) const;
 
+    MAV_DISTANCE_SENSOR get_sensor_type(uint8_t instance) const;
```
in fact it is require otherwise we cannot access to the backend variable

khancyr replied:
orient method added

khancyr commented on libraries/AP_RangeFinder/AP_RangeFinder_uLanding.cpp:
```diff
@@ -28,7 +28,7 @@ extern const AP_HAL::HAL& hal;
                                                              RangeFinder::RangeFinder_State &_state,
                                                              AP_SerialManager &serial_manager) :
-    AP_RangeFinder_Backend(_ranger, instance, _state)
+    AP_RangeFinder_Backend(_ranger, instance, _state, MAV_DISTANCE_SENSOR_LASER)
```
ok I see the point !

khancyr commented on ArduCopter/GCS_Mavlink.cpp:
```diff
@@ -278,9 +278,9 @@ void NOINLINE Copter::send_proximity(mavlink_channel_t chan, uint16_t count_max)
             (uint16_t)(dist_up * 100),      // current distance reading
             MAV_DISTANCE_SENSOR_LASER,      // type from MAV_DISTANCE_SENSOR enum
-            0,                              // onboard ID of the sensor
+            horiz_count + 1,                // onboard ID of the sensor
```
done

khancyr commented:
rebase, add get_sensor_type_orient method, check payload size on send_distance_sensor(),  add send_distance_sensor_downward() as commun helper, add send_rangefinder() as commun function, scale proximity sensors id with new variable : proximity id start at 10, proximity upward is 19

rmackay9 commented:
@khancyr, yes, that may be the best we can do for now at least.

OXINARF commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -219,6 +219,56 @@ bool GCS_MAVLINK::send_battery_status(const AP_BattMonitor &battery) const
+                rangefinder.max_distance_cm(instance),  // maximum distance the sensor can measure in centimeters
+                rangefinder.distance_cm(instance),      // current distance reading (in cm?)
+                rangefinder.get_sensor_type(instance),  // type from MAV_DISTANCE_SENSOR enum
+                instance,                               // onboard ID of the sensor == instance ?
```
Remove the `?` from the comment

khancyr replied:
ok

OXINARF commented on APMrover2/GCS_Mavlink.cpp:
```diff
@@ -417,6 +417,8 @@ bool GCS_MAVLINK_Rover::try_send_message(enum ap_message id)
     case MSG_RANGEFINDER:
         CHECK_PAYLOAD_SIZE(RANGEFINDER);
         rover.send_rangefinder(chan);
+        CHECK_PAYLOAD_SIZE(DISTANCE_SENSOR);
```
You can remove this check, since the next one will check for each message it tries to send.

khancyr replied:
done

OXINARF commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -219,6 +219,56 @@ bool GCS_MAVLINK::send_battery_status(const AP_BattMonitor &battery) const
+    send_distance_sensor(rangefinder, instance);
+}
+
+void GCS_MAVLINK::send_rangefinder(const RangeFinder &rangefinder) const
```
Should this be called *send_rangefinder_downward*?

khancyr replied:
yep, the naming would be clearer!

khancyr commented on libraries/AP_Proximity/AP_Proximity.h:
```diff
@@ -24,6 +24,8 @@
 #define PROXIMITY_YAW_CORRECTION_DEFAULT    22  // default correction for sensor error in yaw
 #define PROXIMITY_MAX_IGNORE                6   // up to six areas can be ignored
+static const int8_t PROXIMITY_MAX_DIRECTION = 8;
+static const int8_t PROXIMITY_SENSOR_ID_START = 10;
```
busted ^^

khancyr replied:
ok changed

khancyr commented:
@OXINARF rebased and change made.
I agree with the CHECK_PAYLOAD_SIZE thing, but that outside the scope of this PR.

I agree that MAV_DISTANCE_SENSOR_ENUM_END isn't good. I will try to make a pr on mavlink enum to have MAV_DISTANCE_SENSOR_UNKNOW. Should we wait it to be merge ?

OXINARF commented:
Yes, I think this should wait on MAVLink change - the more I look at the ENUM_END use, the less I like it :smile: Since you are going to add it, add a radar option too.

khancyr commented:
OK PR submit to mavlink : https://github.com/ArduPilot/mavlink/pull/46

OXINARF commented:
OK, rebased, updated MAVLink submodule to add new rangefinder types, changed code to use them, added type to newly added rangefinder driver, corrected some typos in commit messages and merged! Thanks @khancyr!
